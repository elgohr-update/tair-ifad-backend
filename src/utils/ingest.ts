import parse from "csv-parse/lib/sync";
import { readFileSync } from "fs";
import {evidenceCodes} from "../config";
import {resolve} from "path";

const ANNOTATION_COLUMNS = [
    null,
    "DatabaseID",
    null,
    "Invert",
    "GOTerm",
    "Reference",
    "EvidenceCode",
    "AdditionalEvidence",
    "Aspect",
    "UniqueGeneName",
    "AlternativeGeneName",
    "GeneProductType",
    null,
    "Date",
    "AssignedBy",
    null,
    null,
];

const GENE_COLUMNS = [
    "GeneID",
    "GeneProductType",
];

export type Aspect = "F" | "P" | "C";
export type AnnotationStatus = "KNOWN_EXP" | "KNOWN_OTHER" | "UNKNOWN" | "UNANNOTATED";

export interface IAnnotation {
    DatabaseID: string;
    Invert: boolean;
    GOTerm: string;
    Reference: string;
    EvidenceCode: string;
    AdditionalEvidence: string[];
    Aspect: Aspect;
    AnnotationStatus: AnnotationStatus;
    UniqueGeneName: string;
    AlternativeGeneName: string[];
    GeneProductType: string;
    Date: string;
    AssignedBy: string;
}

export interface IGene {
    GeneID: string;
    GeneProductType: string;
}

export const parseAnnotations = (input: string): IAnnotation[] => {
    return parse(input, {
        columns: ANNOTATION_COLUMNS,
        skip_empty_lines: true,
        delimiter: "\t",
        relax: true,
        cast: (value, context) => {
            if (context.column === "Invert") {
                return value === "NOT";
            } else if (context.column === "AlternativeGeneName" || context.column === "AdditionalEvidence") {
                return value.split("|");
            } else if (context.column === "Date") {
                return new Date(Date.parse(`${value.slice(0,4)}-${value.slice(4,6)}-${value.slice(6,8)}`));
            } else {
                return value;
            }
        },
    })
    .map(item => ({
        ...item,
        AnnotationStatus: evidenceCodeToAnnotationStatus(item.EvidenceCode)
    }));
};

const evidenceCodeToAnnotationStatus = (evidenceCode: string): AnnotationStatus => {
    if (evidenceCodes.UNKNOWN.includes(evidenceCode)) {
        return "UNKNOWN";
    }
    if (evidenceCodes.KNOWN_EXPERIMENTAL.includes(evidenceCode)) {
        return "KNOWN_EXP";
    }
    return "KNOWN_OTHER";
};

export const parseGenes = (input: string): IGene[] => {
    return parse(input, {
        columns: GENE_COLUMNS,
        skip_empty_lines: true,
        skip_lines_with_error: true,
        delimiter: "\t",
        relax: true,
        cast: true,
    });
};

export const readAnnotations = (filename: string) => {
    const buffer = readFileSync(filename).toString();
    const trimmed = buffer.slice(buffer.indexOf("\n"));
    return parseAnnotations(trimmed);
};

export const readGenes = (filename: string) => {
    const buffer = readFileSync(filename).toString();
    const trimmed = buffer.slice(buffer.indexOf("\n"));
    return parseGenes(trimmed);
};

/**
 * An IngestConfig contains the paths of files at which to read and parse data.
 */
export interface IngestConfig {
    genesFile: string;
    annotationsFile: string;
}

/**
 * The default IngestConfig reads filenames from Environment Variables or defaults
 * to the file paths given here.
 */
const defaultConfig: IngestConfig = {
    genesFile: process.env["GENES_FILE"] || resolve("src/assets/gene-types.txt"),
    annotationsFile: process.env["ANNOTATIONS_FILE"] || resolve("src/assets/gene_association.tair"),
};

/**
 * A GroupedAnnotations object contains data which is grouped by the Aspect and the
 * AnnotationStatus of a piece of data. This is generic over any type so that we can
 * build structures of data which may represent different goals. For example, a
 * GroupedAnnotation<Set<IGene>> can be used to organize sets of IGenes that belong to
 * the given Aspect and AnnotationStatus, whereas a GroupedAnnotation<number> can be
 * used to hold the count of how many genes belong to a given Aspect and AnnotationStatus.
 */
export type GroupedAnnotations<T> = {
  [key in Aspect]: {
    all: T,
    unknown: T,
    known: {
      all: T,
      exp: T,
      other: T,
    },
  }
};

/**
 * This function creates a GroupedAnnotations object which is initialized such that every
 * key contains a specific initial value. This can be used to create a GroupedAnnotations
 * which holds empty Sets or which has a 0-count.
 *
 * @param initial A function which produces an initial value to put at each key.
 */
export const makeGroupedAnnotations = <T>(initial: () => T): GroupedAnnotations<T> => ({
  P: { all: initial(), unknown: initial(), known: { all: initial(), exp: initial(), other: initial() } },
  F: { all: initial(), unknown: initial(), known: { all: initial(), exp: initial(), other: initial() } },
  C: { all: initial(), unknown: initial(), known: { all: initial(), exp: initial(), other: initial() } },
});

/**
 * The GeneMap type is an object which is keyed by the names of IGenes and
 * has values of the IGene which the key represents.
 */
export type GeneMap = {
  [key: string]: {
    gene: IGene,
    annotations: Set<IAnnotation>,
  }
};

/**
 * IngestedData is the data format which contains all parsed and cached data
 * which was read from all of the data sources during the ingestion step.
 */
export interface IngestedData {
    geneMap: GeneMap;
    annotations: IAnnotation[];
    groupedAnnotations: GroupedAnnotations<Set<IGene>>;
}

/**
 * The easy top-level entrypoint for reading data needed for backend queries.
 *
 * This function takes an IngestConfig which describes the location for reading
 * files with the data we're interested in. A default configuration is provided
 * which reads these file paths from environment variables.
 *
 * This function returns an `IngestedData` object, which has the parsed data
 * inside of it. The same data may be included in the IngestedData object multiple
 * times if it is transformed or cached in various formats. For example, we have
 * an array of all annotations, or a map in which the annotations are grouped by
 * their aspect and annotation status.
 *
 * @param userConfig An optional configuration for specifying file locations.
 * @return An IngestedData object with parsed and structured information.
 */
export const readData = (userConfig: IngestConfig = defaultConfig): IngestedData => {
    const config: IngestConfig = { ...defaultConfig, ...userConfig };
    console.log("Reading data using configs: " + JSON.stringify(config));

    const genesArray = readGenes(config.genesFile);
    const geneMap: GeneMap = genesArray
        .reduce((acc, current) => {
            acc[current.GeneID] = current;
            return acc;
        }, {});

    // TODO fix parsing so that headers aren't included to start with
    delete geneMap["name"]; // Get rid of "name" header

    const annotations = readAnnotations(config.annotationsFile);

    // Group all annotations based on aspect, categorizing KNOWN_EXP but not KNOWN_OTHER
    const tmpGroupedAnnotations: GroupedAnnotations<Set<IGene>> = annotations
        .reduce((acc, annotation) => {
            const aspect = annotation.Aspect;
            const annotationStatus = annotation.AnnotationStatus;
            const geneId = annotation.AlternativeGeneName.find(name => !!geneMap[name]);
            if (!geneId) return acc;

            const { gene, annotations } = geneMap[geneId];

            // Add the current annotation to the list for the gene it belongs to.
            annotations.add(annotation);

            // Add gene to the proper index in GroupedAnnotations
            acc[aspect].all.add(gene);
            if (annotationStatus === "UNKNOWN") {
                acc[aspect].unknown.add(gene);
            } else {
                acc[aspect].known.all.add(gene);
                if (annotationStatus === "KNOWN_EXP") {
                    acc[aspect].known.exp.add(gene);
                }
            }
            return acc;
        }, makeGroupedAnnotations<Set<IGene>>(() => new Set()));

    // Calculate KNOWN_OTHER by finding all genes that are in "known" but not KNOWN_EXP
    const groupedAnnotations = Object.entries(tmpGroupedAnnotations)
        .reduce((acc, [aspect, index]) => {
              index.known.other = new Set([...index.known.all].filter(gene => !index.known.exp.has(gene)));
              acc[aspect] = index;
              return acc;
        }, makeGroupedAnnotations<Set<IGene>>(() => new Set()));

    return { geneMap, annotations, groupedAnnotations };
};
