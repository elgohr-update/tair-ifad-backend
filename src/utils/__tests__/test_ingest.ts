import {parseAnnotations, parseGenes} from "../ingest";

const gene_data = `
AT1G01010	protein_coding
AT1G01020	protein_coding
AT1G01030	protein_coding
AT1G01040	protein_coding
AT1G01046	miRNA_primary_transcript
AT1G01050	protein_coding
AT1G01060	protein_coding
AT1G01070	protein_coding
AT1G01080	protein_coding
AT1G01090	protein_coding
`;

const annotation_data = `
TAIR	locus:2079747	ALY2		GO:0000003	TAIR:Communication:501741973	IBA	PANTHER:PTN000495591|WB:WBGene00002998	P	AT3G05380	AT3G05380|ALY2|ATALY2|ALWAYS EARLY 2|ARABIDOPSIS THALIANA ALWAYS EARLY 2|F22F7.18|F22F7_18	protein	taxon:3702	20180615	GOC		TAIR:locus:2079747
TAIR	locus:2089438	ALY3		GO:0000003	TAIR:Communication:501741973	IBA	PANTHER:PTN000495591|WB:WBGene00002998	P	AT3G21430	AT3G21430|ATALY3|ALY3|ARABIDOPSIS THALIANA ALWAYS EARLY 3|ALWAYS EARLY 3	protein	taxon:3702	20180615	GOC		TAIR:locus:2089438
TAIR	locus:2143666	ALY1		GO:0000003	TAIR:Communication:501741973	IBA	PANTHER:PTN000495591|WB:WBGene00002998	P	AT5G27610	AT5G27610|ATALY1|ALY1|ARABIDOPSIS THALIANA ALWAYS EARLY 1|ALWAYS EARLY 1|F15A18.70|F15A18_70	protein	taxon:3702	20180615	GOC		TAIR:locus:2143666
TAIR	locus:2179122	APTX		GO:0000012	TAIR:Communication:501741973	IBA	PANTHER:PTN000281062|MGI:MGI:1913658|UniProtKB:Q7Z2E3	P	AT5G01310	AT5G01310|APTX|APRATAXIN-like|T10O8.20|T10O8_20	protein	taxon:3702	20180803	GOC		TAIR:locus:2179122
TAIR	locus:2150931	TDP1		GO:0000012	TAIR:Communication:501741973	IBA	PANTHER:PTN000275870|MGI:MGI:1920036|UniProtKB:Q9NUW8	P	AT5G15170	AT5G15170|TDP1|tyrosyl-DNA phosphodiesterase 1|F8M21.60|F8M21_60	protein	taxon:3702	20180803	GOC		TAIR:locus:2150931
TAIR	locus:2202114	BFN1		GO:0000014	TAIR:Communication:501741973	IBA	PANTHER:PTN002107734|TAIR:locus:2202114	F	AT1G11190	AT1G11190|BFN1|ENDO1|bifunctional nuclease i|ENDONUCLEASE 1|T28P6.14|T28P6_14	protein	taxon:3702	20180615	GOC		TAIR:locus:2202114
TAIR	locus:2045437	RAD50		GO:0000014	TAIR:Communication:501741973	IBA	PANTHER:PTN000429848|UniProtKB:Q92878	F	AT2G31970	AT2G31970|RAD50|ATRAD50|F22D22.28|F22D22_28|dna repair-recombination protein	protein	taxon:3702	20180630	GOC		TAIR:locus:2045437
TAIR	locus:2096329	ERCC1		GO:0000014	TAIR:Communication:501741973	IBA	PANTHER:PTN000297520|UniProtKB:P07992|SGD:S000004560	F	AT3G05210	AT3G05210|ERCC1|UVR7|UV REPAIR DEFICIENT 7|T12H1.18|T12H1_18	protein	taxon:3702	20180803	GOC		TAIR:locus:2096329
TAIR	locus:1005716561	GR1		GO:0000014	TAIR:Communication:501741973	IBA	PANTHER:PTN000387514|UniProtKB:Q99708	F	AT3G52115	AT3G52115|ATGR1|ATCOM1|COM1|GR1|gamma response gene 1	protein	taxon:3702	20180615	GOC		TAIR:locus:1005716561
TAIR	locus:2163011	UVH1		GO:0000014	TAIR:Communication:501741973	IBA	PANTHER:PTN000015911|UniProtKB:Q9LKI5|SGD:S000005943	F	AT5G41150	AT5G41150|UVH1|ATRAD1|RAD1|ULTRAVIOLET HYPERSENSITIVE 1|MEE6.22|MEE6_22|repair endonuclease	protein	taxon:3702	20180615	GOC		TAIR:locus:2163011
`;

describe("Ingestion functions", () => {

    it("should parse a set of Annotations", () => {
        expect(parseAnnotations(annotation_data)).toMatchSnapshot();
    });

    it("should parse a set of Gene IDs and Gene Product Types", () => {
        expect(parseGenes(gene_data)).toMatchSnapshot();
    });

    it("should not fail if the input text is empty", () => {
        expect(() => parseGenes("")).not.toThrow();
    });
});
