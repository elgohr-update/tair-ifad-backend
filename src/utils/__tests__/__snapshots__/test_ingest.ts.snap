// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ingestion functions should parse a set of Annotations 1`] = `
Array [
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000495591",
      "WB:WBGene00002998",
    ],
    "AlternativeGeneName": Array [
      "AT3G05380",
      "ALY2",
      "ATALY2",
      "ALWAYS EARLY 2",
      "ARABIDOPSIS THALIANA ALWAYS EARLY 2",
      "F22F7.18",
      "F22F7_18",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "P",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2079747",
    "Date": 2018-06-15T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000003",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT3G05380",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000495591",
      "WB:WBGene00002998",
    ],
    "AlternativeGeneName": Array [
      "AT3G21430",
      "ATALY3",
      "ALY3",
      "ARABIDOPSIS THALIANA ALWAYS EARLY 3",
      "ALWAYS EARLY 3",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "P",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2089438",
    "Date": 2018-06-15T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000003",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT3G21430",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000495591",
      "WB:WBGene00002998",
    ],
    "AlternativeGeneName": Array [
      "AT5G27610",
      "ATALY1",
      "ALY1",
      "ARABIDOPSIS THALIANA ALWAYS EARLY 1",
      "ALWAYS EARLY 1",
      "F15A18.70",
      "F15A18_70",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "P",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2143666",
    "Date": 2018-06-15T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000003",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT5G27610",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000281062",
      "MGI:MGI:1913658",
      "UniProtKB:Q7Z2E3",
    ],
    "AlternativeGeneName": Array [
      "AT5G01310",
      "APTX",
      "APRATAXIN-like",
      "T10O8.20",
      "T10O8_20",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "P",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2179122",
    "Date": 2018-08-03T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000012",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT5G01310",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000275870",
      "MGI:MGI:1920036",
      "UniProtKB:Q9NUW8",
    ],
    "AlternativeGeneName": Array [
      "AT5G15170",
      "TDP1",
      "tyrosyl-DNA phosphodiesterase 1",
      "F8M21.60",
      "F8M21_60",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "P",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2150931",
    "Date": 2018-08-03T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000012",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT5G15170",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN002107734",
      "TAIR:locus:2202114",
    ],
    "AlternativeGeneName": Array [
      "AT1G11190",
      "BFN1",
      "ENDO1",
      "bifunctional nuclease i",
      "ENDONUCLEASE 1",
      "T28P6.14",
      "T28P6_14",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "F",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2202114",
    "Date": 2018-06-15T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000014",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT1G11190",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000429848",
      "UniProtKB:Q92878",
    ],
    "AlternativeGeneName": Array [
      "AT2G31970",
      "RAD50",
      "ATRAD50",
      "F22D22.28",
      "F22D22_28",
      "dna repair-recombination protein",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "F",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2045437",
    "Date": 2018-06-30T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000014",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT2G31970",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000297520",
      "UniProtKB:P07992",
      "SGD:S000004560",
    ],
    "AlternativeGeneName": Array [
      "AT3G05210",
      "ERCC1",
      "UVR7",
      "UV REPAIR DEFICIENT 7",
      "T12H1.18",
      "T12H1_18",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "F",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2096329",
    "Date": 2018-08-03T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000014",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT3G05210",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000387514",
      "UniProtKB:Q99708",
    ],
    "AlternativeGeneName": Array [
      "AT3G52115",
      "ATGR1",
      "ATCOM1",
      "COM1",
      "GR1",
      "gamma response gene 1",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "F",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:1005716561",
    "Date": 2018-06-15T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000014",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT3G52115",
  },
  Object {
    "AdditionalEvidence": Array [
      "PANTHER:PTN000015911",
      "UniProtKB:Q9LKI5",
      "SGD:S000005943",
    ],
    "AlternativeGeneName": Array [
      "AT5G41150",
      "UVH1",
      "ATRAD1",
      "RAD1",
      "ULTRAVIOLET HYPERSENSITIVE 1",
      "MEE6.22",
      "MEE6_22",
      "repair endonuclease",
    ],
    "AnnotationStatus": "OTHER",
    "Aspect": "F",
    "AssignedBy": "GOC",
    "DatabaseID": "locus:2163011",
    "Date": 2018-06-15T00:00:00.000Z,
    "EvidenceCode": "IBA",
    "GOTerm": "GO:0000014",
    "GeneProductType": "protein",
    "Invert": false,
    "Reference": "TAIR:Communication:501741973",
    "UniqueGeneName": "AT5G41150",
  },
]
`;

exports[`Ingestion functions should parse a set of Gene IDs and Gene Product Types 1`] = `
Array [
  Object {
    "GeneID": "AT1G01010",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01020",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01030",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01040",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01046",
    "GeneProductType": "miRNA_primary_transcript",
  },
  Object {
    "GeneID": "AT1G01050",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01060",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01070",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01080",
    "GeneProductType": "protein_coding",
  },
  Object {
    "GeneID": "AT1G01090",
    "GeneProductType": "protein_coding",
  },
]
`;
